src/Domain/DoctrineDonationRepository.php

```diff
@@ -566,9 +547,11 @@ private function safelyAllocateFunds(Donation $donation, array $fundings, string
             $currentFundingIndex++;
         }
 
-        $this->queueForPersist($donation);
+        if ($withdrewAnything) {
+            $this->queueForPersist($donation);
+            $this->matchingAdapter->saveFundingsToDatabase();
```

Comment:
similarly I'm not sure `saveFundingsToDatabase` is ever doing anything useful. It only does `persist` but the CampaignFundings are not new at this point so surely they're already persisted? We only construct a new CampaignFunding object in `FundRepository::pullForCampaign`

Reply:
(to confirm we don't appear to use the [Deferred Explicit](https://www.doctrine-project.org/projects/doctrine-orm/en/3.3/reference/change-tracking-policies.html#deferred-explicit) strategy anywhere - we use the default Deferred Implicit. 

Which means we don't have to ever call `persist` on an already doctrine-managed entity - its only needed for newly instantiated entities.

Reply:
Yeah this does sound right – it would be great news if so as it's a commonly used and currently expensive fn. I'll look more closely later and double check that everything calling it has a flush after, as the current approach looks like it might well be a side effect of me misunderstanding this a while ago.

Reply:
Right. Afaik calling persist on an object that's doctrine-managed is useless whether or not there's a `flush` anywhere. Calling flush is the only bit that matters.

Reply:
If `Matching\Adapter::saveFundingsToDatabase` is useless as it seems to be then `Matching\Adapter::addAmountWithoutSavingFundingsToDB` is misleadingly named - it's always without saving to DB until we do a 'flush'.

Reply:
Yeah, I was thinking about inlining as the next step. It looks like this got renamed from addAmountWithoutNewTransaction() https://github.com/thebiggive/matchbot/commit/ff2a8d7d025771afe3d14fd6c6644288bf6fbd83 and maybe the goal was to emphasise that we should flush too? But I don't think either name really makes this clearer or is likely to be preventing bugs.

Reply:
I think the confusion has been in the code for a long time - at least since 2019 when we were calling `runTransactionally` but passing it a a closure that afaik didn't actually do an database work:

https://github.com/thebiggive/matchbot/blob/733d1db09f872bebd2cea29c22442ad3c6d4667b/src/Domain/DonationRepository.php#L115

Reply:
@bdsl I think back then we were trying out Redis for the first time and still had Doctrine as an option for the primary matching adapter. So probably the transaction made sense, kind of, depending on which one was in service. Then we kept building on it and forgot to simplify when it became Redis-only.